repo/content/index/index.go

```diff
@@ -26,33 +25,40 @@ type Index interface {
 }
 
 // Open reads an Index from a given reader. The caller must call Close() when the index is no longer used.
-func Open(readerAt io.ReaderAt, v1PerContentOverhead uint32) (Index, error) {
-	h, err := v1ReadHeader(readerAt)
+func Open(data []byte, closer func() error, v1PerContentOverhead uint32) (Index, error) {
+	h, err := v1ReadHeader(data)
 	if err != nil {
 		return nil, errors.Wrap(err, "invalid header")
 	}
 
 	switch h.version {
 	case Version1:
-		return openV1PackIndex(h, readerAt, v1PerContentOverhead)
+		return openV1PackIndex(h, data, closer, v1PerContentOverhead)
 
 	case Version2:
-		return openV2PackIndex(readerAt)
+		return openV2PackIndex(data, closer)
 
 	default:
 		return nil, errors.Errorf("invalid header format: %v", h.version)
 	}
 }
 
-func readAtAll(ra io.ReaderAt, p []byte, offset int64) error {
-	n, err := ra.ReadAt(p, offset)
-	if n != len(p) {
-		return errors.Errorf("incomplete read at offset %v, got %v bytes, expected %v", offset, n, len(p))
-	}
+func safeSlice(data []byte, offset int64, length int) (v []byte, err error) {
+	defer func() {
+		if recover() != nil {
+			err = errors.Errorf("invalid slice offset=%v, length=%v, data=%v bytes", offset, length, len(data))
```

Comment:
Explicitly return a `nil []byte`  ?

```suggestion
			v = nil
			err = errors.Errorf("invalid slice offset=%v, length=%v, data=%v bytes", offset, length, len(data))
```